src/cpp/rtps/builtin/discovery/endpoint/EDPStatic.cpp

```diff
@@ -251,6 +259,209 @@ std::pair<std::string, std::string> EDPStaticProperty::toProperty(
+    }
+    return out;
+}
+
+static std::string base64_decode(
+        const std::string& in)
+{
+
+    std::string out;
+
+    std::vector<int> T(256, -1);
+    for (int i = 0; i < 64; i++)
+    {
+        T[base64_str[i]] = i;
+    }
+
+    int val = 0, valb = -8;
+    for (uchar c : in)
+    {
+        if (T[c] == -1)
+        {
+            break;
+        }
+        val = (val << 6) + T[c];
+        valb += 6;
+        if (valb >= 0)
+        {
+            out.push_back(char((val >> valb) & 0xFF));
+            valb -= 8;
+        }
+    }
+    return out;
+}
+
+std::string vector_to_string(
+        const std::vector<uint8_t>& bits)
+{
+    assert(0 == bits.size() % 8);
+    std::string bits_str;
+
+    for (size_t i {0}; i < bits.size(); i += 8)
+    {
+        uint8_t byte {0};
+        for (size_t j {0}; j < 8; ++j)
+        {
+            byte |= bits[i + j] << (7 - j);
+        }
+        bits_str.push_back(byte);
+    }
+
+    return base64_encode(bits_str);
+}
+
+std::vector<uint8_t> string_to_vector(
+        const std::string& bits)
+{
+    std::vector<uint8_t> bits_vector;
+    std::string bits_str {base64_decode(bits)};
+
+    for (uint8_t byte : bits_str)
+    {
+        for (size_t i {0}; i < 8; ++i)
+        {
+            bits_vector.push_back(byte >> (7 - i) & 0x01);
+        }
+    }
+
+    return bits_vector;
+}
+
+bool EDPStatic::enable_reader_on_v2_property(
+        const std::string& local_participant_name,
+        fastdds::dds::ParameterPropertyList_t& pdp_properties,
+        uint16_t id,
+        bool disable)
+{
+    bool ret_value {false};
+    auto property_it = std::find_if(pdp_properties.begin(), pdp_properties.end(),
+                    [](const fastdds::dds::ParameterProperty_t& property)
+                    {
+                        if (0 == property.first().compare("ESR"))
+                        {
+                            return true;
+                        }
+
+                        return false;
+                    });
+
+    std::vector<uint8_t> bits;
+    std::pair<std::string, std::string> prop;
+
+    if (pdp_properties.end() == property_it)
+    {
+        // Create empty property.
+        prop.first = "ESR";
+        size_t number_of_readers = mp_edpXML->get_number_of_readers(local_participant_name);
+        // Make multiple of 8 bits
+        number_of_readers = (number_of_readers + 7) & ~7;
+        bits = std::vector<uint8_t>(number_of_readers, 0);
+        prop.second = vector_to_string(bits);
+        pdp_properties.push_back(prop);
+        property_it = pdp_properties.begin();
+        auto next_it = pdp_properties.begin();
+        while (++next_it != pdp_properties.end())
+        {
+            property_it = next_it;
+        }
+    }
+    else
+    {
+        bits = string_to_vector(property_it->second());
+    }
+
+    uint32_t position {0};
+    if (xmlparser::XMLP_ret::XML_OK == mp_edpXML->lookforReader(local_participant_name.c_str(), id, nullptr, position))
```

Comment:
I would do this lookup first, and do an early return if the reader is not found. That way we avoid the other (expensive) operations when not needed.